proposals/003-authentication-api.md

```diff
@@ -0,0 +1,650 @@
+
+```java
+    /**
+     * Returns the SASL context for the client connection, or empty if the client
+     * has not successfully authenticated using SASL.
+     * Filters should use {@link #clientPrincipal()} in preference to this method, unless they require SASL-specific functionality.
+     * @return The SASL context for the client connection, or empty if the client
+     * has not successfully authenticated using SASL.
+     */
+    Optional<ClientSaslContext> clientSaslContext();
+```
+
+Where
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+import java.util.Optional;
+
+import io.kroxylicious.proxy.filter.FilterContext;
+
+/**
+ * Exposes SASL authentication information to plugins, for example using {@link FilterContext#clientSaslContext()}.
+ * This is implemented by the runtime for use by plugins.
+ */
+public interface ClientSaslContext {
+
+    /**
+     * The name of the SASL mechanism used by the client.
+     * @return The name of the SASL mechanism used by the client.
+     */
+    String mechanismName();
+
+    /**
+     * Returns the client's authorizationId that resulted from the SASL exchange.
+     * @return the client's authorizationId.
+     */
+    String authorizationId();
+
+    /**
+     * The server identity that the proxy presented to the client using SASL authentication.
+     * @return the proxy's identity with the client. This will be null
+     * if the proxy did not supply an identity because the SASL mechanism used
+     * does not support mutual authentication.
+     */
+    Optional<String> proxyServerId();
+}
+```
+
+### APIs for using client principals in a generic way
+
+Most `Filters` don't need to be opinionated about how the client is identified. They only need to know:
+
+* that the client _is_ authenticated, somehow
+* the name of the client's identity
+
+We want to avoid making `Filter` developers pick a source of authentication information (`clientTlsContext()` or `clientSaslContext()`) in order to maximise the reusabilty of `Filter` implementations. 
+For this purpose we will make use of `java.security.Principal` by providing the following method on `FilterContext`:
+
+```java
+    /**
+     * <p>Returns the authenticated principal for the client connection.</p>
+     * 
+     * <p>The concrete type of principal returned depends on the proxy configuration.
+     * For example,
+     * it may be a {@link javax.security.auth.x500.X500Principal} if client identity is TLS-based,
+     * or it may be a {@link SaslPrincipal} if client identity is SASL based.</p>
+     *
+     * <p>Callers should not:</p>
+     * <ul>
+     *   <li>assume any particular type of principal.
+     *   <li>assume the principal does not change during the lifetime of a filter (due to reauthentication)
+     * </ul>
+     *
+     * @return The authenticated principal for the client connection, or empty if the client
+     * has not successfully authenticated.
```

Comment:
What does this return when a client presents a client TLS cert and then something performs SASL initiation? As presumably the person configuring the proxy would have configured `SASL` as the principal type?

On that note I guess a plugin wanting to use the client TLS cert to initiate SASL would need to use the `clientTlsContext` rather than the principal. 

The timing dependent nature of this leaves me a feeling a bit concerned. I need to mull it further.

Reply:
maybe we could allow filters to specify their own principalType in the filterDefinition? By default they'd receive the virtual cluster principalType, but could override it. 

You would have to amend the startup validations, since a VirtualCluster with principalType=SASL containing a filter that has specified principalType=X500 would need to ensure mTLS is enabled and SASL components are in the filter chain. Plus you'd want to validate that a filterfactory that consumes X509Principals can't declare it wants SASL principalType.